probably/pgcl/check.py

```diff
@@ -198,8 +198,23 @@ def get_type(program: Program,
 
 
 def _check_distribution_arguments(prog: Program, expected_type: Type, *parameters: Tuple[Type, Expr]) -> Union[Type, CheckFail]:
+    # this function is copied from the syntax module, which is not pretty, but prevents cyclic imports
+    # also, it works only if constant substitution has been performed
+    # pylint: disable=duplicate-code
+    def _has_variable(expr: Expr) -> bool:
```

Comment:
There must be a prettier way to do this, right? The only way I can think of is to put `has_variable` into the typechecker module, which doesn't seem right either

Reply:
The reason `syntax` depends on `check` is that it just imports `CheckFail`. That's just a class to report errors on the AST. Meanwhile, `check` is mostly about typechecking. I think we should split the `check` module into two modules: `tycheck` (to distinguish from `syntax`, which also does some checks), and `errors`/`diagnostic` for error and warning infrastructure.

Reply:
Also, where does the cyclic import even come from?

Reply:
> Also, where does the cyclic import even come from?

If we import `has_variables` from the syntax module, which again imports `CheckFail` from the typechecking module

Reply:
Resolved, because this check has been removed